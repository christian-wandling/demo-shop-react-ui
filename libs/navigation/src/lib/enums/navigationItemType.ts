/**
 * Enum defining the types of navigation items in the application
 */
export enum NavigationItemType {
  /**
   * Represents a navigational item that is a route
   */
  ROUTE = 'ROUTE',
}
