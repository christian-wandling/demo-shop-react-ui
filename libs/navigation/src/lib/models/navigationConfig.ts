import { NavigationItem } from './navigationItem';

export interface NavigationConfig {
  items: NavigationItem[];
}
