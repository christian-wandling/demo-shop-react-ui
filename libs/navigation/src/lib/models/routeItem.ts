import { NavigationItemType } from '../enums/navigationItemType';
import { NavigationItem } from './navigationItem';
import { PermissionStrategy } from '@demo-shop-react-ui/auth';

/**
 * Represents a navigation item that links to an Angular route
 */
export class RouteItem extends NavigationItem {
  /**
   * The type of navigation item, always set to ROUTE for RouteItem
   */
  override readonly type: NavigationItemType = NavigationItemType.ROUTE;

  constructor(
    label: string,
    order: number,
    public override options: {
      route: string;
      permissionStrategy?: PermissionStrategy;
      query?: string;
    },
    subItems?: RouteItem[]
  ) {
    super(label, order, options, subItems);
  }
}
