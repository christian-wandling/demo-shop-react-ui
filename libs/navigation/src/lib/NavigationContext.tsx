import { createContext } from 'react';
import { NavigationConfig } from './models/navigationConfig';

export const NavigationContext = createContext<NavigationConfig | null>(null);
