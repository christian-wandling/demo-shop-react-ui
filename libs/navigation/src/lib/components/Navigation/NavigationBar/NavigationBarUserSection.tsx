import { UserNavigationProps } from '../../../models/userNavigationProps';

export const NavigationBarUserSection = ({ user, onLogin, onLogout, onRegister }: UserNavigationProps) => {
  if (user) {
    return (
      <div className="hidden lg:flex lg:flex-1 lg:items-center lg:justify-end lg:space-x-6">
        <a className="text-sm font-medium text-gray-700  cursor-pointer">
          {user.firstname}&nbsp;{user.lastname}
        </a>
        <span className="h-6 w-px bg-gray-200" aria-hidden="true"></span>
        <a onClick={onLogout} className="text-sm font-medium text-gray-700 hover:text-indigo-500 cursor-pointer">
          Sign out
        </a>
      </div>
    );
  }

  return (
    <div className="hidden lg:flex lg:flex-1 lg:items-center lg:justify-end lg:space-x-6">
      <a onClick={onLogin} className="text-sm font-medium text-gray-700 hover:text-indigo-500 cursor-pointer">
        Sign in
      </a>
      <span className="h-6 w-px bg-gray-200" aria-hidden="true"></span>
      <a onClick={onRegister} className="text-sm font-medium text-gray-700 hover:text-indigo-500 cursor-pointer">
        Register
      </a>
    </div>
  );
};
