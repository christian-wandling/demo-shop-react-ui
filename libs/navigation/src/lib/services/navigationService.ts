// Assuming this exists in React form
import { NavigationItem } from '../models/navigationItem';
import { hasPermission, PermissionStrategy } from '@demo-shop-react-ui/auth';
import { NavigationItemType } from '../enums/navigationItemType';

/**
 * Returns navigation items filtered by type and user permissions
 *
 * This method filters the navigation items based on:
 * 1. The specified navigation type
 * 2. The user's permissions (if a permission strategy is defined)
 * The resulting items are sorted by their order property in ascending order.
 */
export function getFilteredNavigationItems(
  items: NavigationItem[],
  type: NavigationItemType = NavigationItemType.ROUTE
): NavigationItem[] {
  const isType = (itemType: NavigationItemType) => itemType === type;
  const allowDisplay = (strategy?: PermissionStrategy) => !strategy || hasPermission(strategy);

  return items
    .filter(item => isType(item.type) && allowDisplay(item.options?.['permissionStrategy']))
    .sort((a, b) => a.order - b.order);
}
