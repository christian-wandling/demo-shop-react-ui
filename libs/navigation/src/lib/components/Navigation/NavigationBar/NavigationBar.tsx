import { Link } from 'react-router';
import { ProductSearch } from '@demo-shop-react-ui/product';
import { CartIcon } from '@demo-shop-react-ui/shopping';
import { NavigationBarProps } from '../../../models/navigationBarProps';
import { RouteItem } from '../../../models/routeItem';
import { UserNavigation } from '../UserNavigation/UserNavigation';

/**
 * NavigationDesktop renders the desktop version of the navigation.
 * It includes the main navigation bar with logo, links, search and cart.
 */
export const NavigationBar = ({
  items,
  selectedItem,
  user,
  onOpenSideNavigation,
  onSelectedItemChange,
  onLogin,
  onRegister,
  onLogout,
}: NavigationBarProps) => {
  return (
    <header className="sticky top-0 z-20 bg-white" data-testid="navigation-bar">
      <nav aria-label="Top" className="mx-auto max-w-7xl px-4 sm:px-6 lg:px-8">
        <div className="border-b border-gray-200">
          <div className="flex h-16 items-center">
            <button
              onClick={onOpenSideNavigation}
              type="button"
              className="relative rounded-md bg-white p-2 text-gray-400 lg:hidden"
              data-testid="navigation-slide-over-toggle">
              <span className="absolute -inset-0.5"></span>
              <span className="sr-only">Open menu</span>
              <svg
                className="h-6 w-6"
                fill="none"
                viewBox="0 0 24 24"
                strokeWidth="1.5"
                stroke="currentColor"
                aria-hidden="true">
                <path strokeLinecap="round" strokeLinejoin="round" d="M3.75 6.75h16.5M3.75 12h16.5m-16.5 5.25h16.5" />
              </svg>
            </button>

            <div className="ml-4 flex lg:ml-0 h-8 w-8 relative" data-testid="navigation-bar-logo">
              <Link to="/">
                <span className="sr-only">Demo Shop</span>
                <img className="object-contain" src="/icons/demo-shop.png" alt="Demo shop logo" />
              </Link>
            </div>

            <div className="hidden lg:ml-8 lg:block lg:self-stretch">
              <div className="flex h-full space-x-8">
                {items.map((item, index) => (
                  <div key={index} className="menu-item flex h-full space-x-8">
                    <Link
                      onClick={() => onSelectedItemChange(item.label)}
                      to={(item as RouteItem).options?.route}
                      className={
                        (selectedItem === item.label ? 'text-indigo-600' : 'text-gray-700') +
                        'menu-item-button flex items-center text-sm font-medium hover:text-indigo-500'
                      }
                      data-testid={`navigation-bar-item-${item.label}`}>
                      {item.label.charAt(0).toUpperCase() + item.label.slice(1)}
                    </Link>
                  </div>
                ))}
              </div>
            </div>

            <div className="ml-auto flex items-center">
              <UserNavigation
                className="hidden lg:flex lg:flex-1 lg:items-center lg:justify-end lg:space-x-6"
                data-testid="navigation-bar-user-section"
                user={user}
                onLogin={onLogin}
                onLogout={onLogout}
                onRegister={onRegister}
              />
              <ProductSearch />
              <CartIcon />
            </div>
          </div>
        </div>
      </nav>
    </header>
  );
};
