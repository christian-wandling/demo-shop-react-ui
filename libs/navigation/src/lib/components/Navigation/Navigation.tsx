import { useContext, useEffect, useState } from 'react';
import { NavigationBar } from './NavigationBar/NavigationBar';
import { SideNavigation } from './SideNavigation/SideNavigation';
import { getFilteredNavigationItems } from '../../services/navigationService';
import { NavigationItemType } from '../../enums/navigationItemType';
import { NavigationContext } from '../../NavigationContext';
import { login, logout, register } from '@demo-shop-react-ui/auth';
import { useUserStore } from '@demo-shop-react-ui/user';

/**
 * NavigationContainer is responsible for managing state and data for the navigation components.
 * It renders different navigation components based on screen size.
 */
export const Navigation = () => {
  const config = useContext(NavigationContext);
  const user = useUserStore(state => state.user);

  const [sideNavigationOpen, setSideNavigationOpen] = useState(false);
  const [selectedNavigationItem, setSelectedNavigationItem] = useState('products');
  const navigationItems = getFilteredNavigationItems(config?.items ?? [], NavigationItemType.ROUTE);

  useEffect(() => {
    const handleResize = () => setSideNavigationOpen(false);
    window.addEventListener('resize', handleResize);
    return () => window.removeEventListener('resize', handleResize);
  }, []);

  /**
   * Handles user login and loads shopping session
   */
  const handleLogin = async () => {
    await login();
    // await cart.loadShoppingSession();
  };

  /**
   * Handles user registration and loads shopping session
   */
  const handleRegister = async () => {
    await register();
    // await cart.loadShoppingSession();
  };

  /**
   * Handles user logout
   */
  const handleLogout = async () => {
    await logout();
  };

  return (
    <div className="bg-white">
      <NavigationBar
        items={navigationItems}
        selectedItem={selectedNavigationItem}
        user={user}
        onOpenSideNavigation={() => setSideNavigationOpen(true)}
        onSelectedItemChange={(item: string) => setSelectedNavigationItem(item)}
        onLogin={handleLogin}
        onRegister={handleRegister}
        onLogout={handleLogout}
      />
      {sideNavigationOpen && (
        <SideNavigation
          items={navigationItems}
          selectedItem={selectedNavigationItem}
          user={user}
          isOpen={sideNavigationOpen}
          onSetOpen={open => setSideNavigationOpen(open)}
          onSelectedItemChange={(item: string) => setSelectedNavigationItem(item)}
          onLogin={handleLogin}
          onRegister={handleRegister}
          onLogout={handleLogout}
        />
      )}
    </div>
  );
};
