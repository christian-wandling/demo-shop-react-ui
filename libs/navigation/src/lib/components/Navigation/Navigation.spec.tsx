import { beforeEach, describe, expect, it, vi } from 'vitest';
import { act, render, screen } from '@testing-library/react';
import { Navigation } from './Navigation';
import { NavigationItemType } from '../../enums/navigationItemType';
import { RouteItem } from '../../models/routeItem';
import { PermissionStrategy } from '@demo-shop-react-ui/auth';
import { NavigationConfig } from '../../models/navigationConfig';
import { NavigationContext } from '../../NavigationContext';
import { NavigationBar } from './NavigationBar/NavigationBar';
import { getFilteredNavigationItems } from '../../services/navigationService';

vi.mock('../../services/navigationService', () => ({
  getFilteredNavigationItems: vi.fn().mockImplementation((items, type) => [
    { id: 'products', label: 'Products', type: NavigationItemType.ROUTE },
    { id: 'categories', label: 'Categories', type: NavigationItemType.ROUTE },
  ]),
}));

const mockShoppingCartState = {
  fetchCurrentSession: vi.fn(),
};

vi.mock('@demo-shop-react-ui/shopping', () => ({
  useShoppingCartStore: vi.fn(selectorOrState => {
    if (typeof selectorOrState === 'function') {
      return selectorOrState(mockShoppingCartState);
    }
    return mockShoppingCartState;
  }),
}));

const login = vi.fn().mockResolvedValue(undefined);
const register = vi.fn().mockResolvedValue(undefined);
const logout = vi.fn().mockResolvedValue(undefined);

vi.mock('@demo-shop-react-ui/auth', async importOriginal => ({
  ...(await importOriginal()),
  login: () => login(),
  register: () => register(),
  logout: () => logout(),
}));

vi.mock('./NavigationBar/NavigationBar', () => ({
  NavigationBar: vi.fn(() => <div data-testid="navigation-bar"></div>),
}));

vi.mock('./SideNavigation/SideNavigation', () => ({
  SideNavigation: vi.fn(() => <div data-testid="side-navigation"></div>),
}));

const navigationConfig: NavigationConfig = {
  items: [
    new RouteItem('products', 101, { route: 'products' }),
    new RouteItem('orders', 102, { route: 'orders', permissionStrategy: PermissionStrategy.AUTHENTICATED }),
  ],
};

describe('Navigation Component', () => {
  const fetchCurrentSessionSpy = vi.spyOn(mockShoppingCartState, 'fetchCurrentSession');

  beforeEach(() => {
    vi.clearAllMocks();
    global.window.resizeTo = vi.fn();
    global.window.addEventListener = vi.fn();
    global.window.removeEventListener = vi.fn();
  });

  it('renders navigation bar', () => {
    render(
      <NavigationContext.Provider value={navigationConfig}>
        <Navigation />
      </NavigationContext.Provider>
    );

    expect(screen.getByTestId('navigation-bar')).toBeTruthy();
    expect(screen.queryByTestId('side-navigation')).not.toBeTruthy();
  });

  it('matches the snapshot', () => {
    const { baseElement } = render(
      <NavigationContext.Provider value={navigationConfig}>
        <Navigation />
      </NavigationContext.Provider>
    );
    expect(baseElement).toMatchSnapshot();
  });

  it('calls getFilteredNavigationItems with correct parameters', () => {
    render(
      <NavigationContext.Provider value={navigationConfig}>
        <Navigation />
      </NavigationContext.Provider>
    );

    expect(getFilteredNavigationItems).toHaveBeenCalledWith(navigationConfig.items, NavigationItemType.ROUTE);
  });

  it('adds and removes resize event listener', () => {
    const { unmount } = render(
      <NavigationContext.Provider value={navigationConfig}>
        <Navigation />
      </NavigationContext.Provider>
    );

    expect(window.addEventListener).toHaveBeenCalledWith('resize', expect.any(Function));

    unmount();

    expect(window.removeEventListener).toHaveBeenCalledWith('resize', expect.any(Function));
  });

  it('opens side navigation when triggered', async () => {
    render(
      <NavigationContext.Provider value={navigationConfig}>
        <Navigation />
      </NavigationContext.Provider>
    );

    const navigationBarProps = NavigationBar.mock.calls[0][0];

    await act(async () => {
      navigationBarProps.onOpenSideNavigation();
    });

    expect(screen.getByTestId('side-navigation')).toBeTruthy();
  });

  it('changes selected navigation item', async () => {
    render(
      <NavigationContext.Provider value={navigationConfig}>
        <Navigation />
      </NavigationContext.Provider>
    );

    const navigationBarProps = NavigationBar.mock.calls[0][0];
    expect(navigationBarProps.selectedItem).toBe('products');

    await act(async () => {
      navigationBarProps.onSelectedItemChange('about');
    });

    const updatedProps = NavigationBar.mock.calls[NavigationBar.mock.calls.length - 1][0];
    expect(updatedProps.selectedItem).toBe('about');
  });

  it('calls login function when triggered', async () => {
    render(
      <NavigationContext.Provider value={navigationConfig}>
        <Navigation />
      </NavigationContext.Provider>
    );

    const navigationBarProps = NavigationBar.mock.calls[0][0];

    await act(async () => {
      await navigationBarProps.onLogin();
    });

    expect(login).toHaveBeenCalledTimes(1);
    expect(fetchCurrentSessionSpy).toHaveBeenCalledTimes(1);
  });

  it('calls register function when triggered', async () => {
    render(
      <NavigationContext.Provider value={navigationConfig}>
        <Navigation />
      </NavigationContext.Provider>
    );

    const navigationBarProps = NavigationBar.mock.calls[0][0];

    await act(async () => {
      await navigationBarProps.onRegister();
    });

    expect(register).toHaveBeenCalledTimes(1);
    expect(fetchCurrentSessionSpy).toHaveBeenCalledTimes(1);
  });

  it('calls logout function when triggered', async () => {
    render(
      <NavigationContext.Provider value={navigationConfig}>
        <Navigation />
      </NavigationContext.Provider>
    );

    const navigationBarProps = NavigationBar.mock.calls[0][0];

    await act(async () => {
      await navigationBarProps.onLogout();
    });

    expect(logout).toHaveBeenCalledTimes(1);
  });

  it('closes side navigation when window resizes', async () => {
    render(
      <NavigationContext.Provider value={navigationConfig}>
        <Navigation />
      </NavigationContext.Provider>
    );

    const resizeHandler = window.addEventListener.mock.calls.find(call => call[0] === 'resize')[1];

    const navigationBarProps = NavigationBar.mock.calls[0][0];

    await act(async () => {
      navigationBarProps.onOpenSideNavigation();
      resizeHandler();
    });

    expect(screen.queryByTestId('side-navigation')).not.toBeTruthy();
  });
});
