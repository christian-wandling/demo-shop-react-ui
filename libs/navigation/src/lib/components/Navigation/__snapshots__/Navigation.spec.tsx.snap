// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Navigation Component > matches the snapshot 1`] = `
<body>
  <div>
    <div
      data-testid="navigation-bar"
    />
  </div>
</body>
`;
